packages/expo-modules-core/common/cpp/EventEmitter.h

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#ifdef __cplusplus
+
+#include <unordered_map>
+#include <list>
+#include <jsi/jsi.h>
+
+namespace jsi = facebook::jsi;
+
+namespace expo::EventEmitter {
+
+/**
+ Class containing and managing listeners of the event emitter.
+ */
+class Listeners {
+private:
+  friend class NativeState;
+  friend void installClass(jsi::Runtime &runtime);
+
+  /**
+   Type of the list containing listeners for the specific event name.
+   */
+  using ListenersList = std::list<jsi::Value>;
+
+  /**
+   Type of the map where the keys are event names and the values are lists of listeners.
+   */
+  using ListenersMap = std::unordered_map<std::string, ListenersList>;
+
+  /**
+   Map with the events and listeners.
+   */
+  ListenersMap listenersMap;
+
+  /**
+   Adds a listener for the given event name.
+   */
+  void add(jsi::Runtime &runtime, std::string eventName, const jsi::Function &listener) noexcept;
+
+  /**
+   Removes the listener for the given event name.
+   */
+  void remove(jsi::Runtime &runtime, std::string eventName, const jsi::Function &listener) noexcept;
+
+  /**
+   Removes all listeners for the given event name.
+   */
+  void removeAll(std::string eventName) noexcept;
+
+  /**
+   Clears the entire map of events and listeners.
+   */
+  void clear() noexcept;
+
+  /**
+   Calls listeners for the given event name, with the given `this` object and payload arguments.
+   */
+  void call(jsi::Runtime &runtime, std::string eventName, const jsi::Object &thisObject, const jsi::Value *args, size_t count) noexcept;
+};
+
+/**
+ Class representing a native state of objects that emit events.
+ */
+class JSI_EXPORT NativeState : public jsi::NativeState {
```

Comment:
```suggestion
class JSI_EXPORT PublisherNativeState : public jsi::NativeState {
```
Could we consider renaming this class to a more descriptive name, such as `PublisherNativeState`? While the class is currently in the `EventEmitter` namespace, it may be beneficial to have this native state associated with something else in the same namespace in the future.